### PR TITLE
Use resumable uploads for large files.

### PR DESCRIPTION
**Pull request by coryan**
With this change UploadFile() uses resumable uploads for large files
(how large is a configuration parameter).

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1550)

### CODE REVIEW

File: google/cloud/storage/client.cc [hunk 1]
```diff
@@ -50,58 +48,97 @@ This is often a problem because:
...
   }
-  auto response = streambuf->Close();
-  if (response.status_code >= 300) {
-    std::ostringstream os;
-    os << __func__ << ": error in during upload "
-       << Status(response.status_code, response.payload);
-    google::cloud::internal::RaiseRuntimeError(os.str());
+
+  // GCS requires chunks to be a multiple of 256KiB.
+  auto chunk_size = internal::UploadChunkRequest::RoundUpToQuantum(
```
Comment (houglum) [inline google/cloud/storage/client.cc]: We can tune this in a future PR, but I want to note that having a chunk size too small will result in sending a silly amount of requests if the object is large.  We use a pretty large value in gsutil by default (100M):
https://github.com/GoogleCloudPlatform/gsutil/blob/88ab2023f0dcf3d7a5444832eb547b2cbc68d7bd/gslib/util.py#L833

File: google/cloud/storage/client.cc [hunk 2]
```diff
@@ -33,9 +33,7 @@ static_assert(std::is_copy_assignable<storage::Client>::value,
 Client::Client(ClientOptions options)
     : Client(internal::CurlClient::Create(std::move(options))) {}
 
-ObjectMetadata Client::UploadFileImpl(
-    std::string const& file_name,
-    internal::InsertObjectStreamingRequest request) {
+bool Client::UseSimpleUpload(std::string const& file_name) const {
   auto status = google::cloud::internal::status(file_name);
   if (not is_regular(status)) {
     GCP_LOG(WARNING) << "Trying to upload " << file_name
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/client.cc]: Fixed.

File: google/cloud/storage/client.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/client.cc]: Ack.

**Comment by coryan** [comment]
PTAL.

**Pull request merged**